### PR TITLE
Fix plugin topological ordering

### PR DESCRIPTION
**Pull request by zachmullen**
Transitive plugin dependencies were broken. This unifies the code
path for DAG traversal and toposorting of plugins between the
storage code in the setting model and the loading code, and now
properly supports transitive dependency ordering. This also dies
gracefully if cycles exist.

### CODE REVIEW

**Comment by zachmullen** [comment]
@mgrauer PTAL to make sure this fixes your issue. In my manual testing it worked for transitive dependencies where the old one failed. If this fix works for you, I'll add some testing for it.

**Comment by zachmullen** [comment]
@mgrauer that last commit should fix the issue for you.

**Comment by mgrauer** [comment]
This solves the issue for me.  It sorts the plugins list twice, but it isn't a big deal and it only happens at server startup.

Looking at `loadPlugins`, it has [this comment](https://github.com/girder/girder/blob/e7b8af984447af76690d8342e6f4b10728206844/girder/utility/plugin_utilities.py#L47-L49):

```
Loads a set of plugins into the application. The list passed in should not
already contain dependency information; dependent plugins will be loaded
automatically.
```

Can you rephrase this comment to explain what `should not
already contain dependency information` means?  Does passing in a topo sorted list violate this?

**Comment by zachmullen** [comment]
@mgrauer PTAL, I updated the comments and also made it so we don't build the DAG and sort it twice on server startup.

**Comment by mgrauer** [comment]
Great, LGTM.

**Comment by danlamanna** [comment]
We might not necessarily care, but this leaves Girder unable to start if a plugin that is expected no longer exists on disk. The only way I can see it being done is by manually altering the setting with mongo.

**Comment by jbeezley** [comment]
Well, I care a little.  What is it that prevents it from starting?

**Comment by zachmullen** [comment]
I'm putting in a fix shortly :)
